### DISCUSSION TITLE
how can I pass arguments in a plugin invoked by AutoInvokeKernelFunctions?

### DISCUSSION THREAD

**Discussion by MithrilMan** [microsoft/semantic-kernel#7977] [Q&A] [answered]
I've a plugin that gets called automatically by using AutoInvokeKernelFunctions.
I don't use any kernel to run the main prompt but I use a resolved IChatCompletionService.

IChatCompletionService accepts a kernel as argument but don't accepts a `KernelArguments` (that I could use to pass the data I want), how am I supposed to pass some arguments to the prompt that may then arrive finally to my plugin if invoked?

To be more specific in this case I've to simply pass a "conversationId"

**Comment by MithrilMan** [comment]
oh, I've find this reply https://github.com/microsoft/semantic-kernel/discussions/6457#discussioncomment-9620779 that could fit my need, btw I think.

Is this property properly documented?

**Reply by dmytrostruk** [reply, marked as answer]
@MithrilMan We have C# API reference, where this property is mentioned, are you interested in any specific information about it? Thanks.
https://learn.microsoft.com/en-us/dotnet/api/microsoft.semantickernel.kernel.data?view=semantic-kernel-dotnet#microsoft-semantickernel-kernel-data

**Reply by MithrilMan** [reply]
@dmytrostruk Yes I mean, that's a super important property that doesn't stand out as it should.
I don't know if it's me that I've missed some samples using it but I think it should be highlighted in the plugin API because it's the main (or let say better) way to pass contextual information when KernelArguments can't be used.
I've seen we can use Filters too, that's more toward AOP way to code.

I think a proper full example on Autofunction call with "data injection" would close the gap of missing information.

I understand that this library is kind of WIP and many things are going to change and cause breaking changes until things settle down a bit in the space, and I understand (I know personally) the cost of documenting rapidly evolving stuff, but the main problem I see on current doc is a comprehensive "cook book" with advanced scenarios: once you know about a concept (e.g. Filters or this Kernel.Data) you can then goole and find info around like your blog post on filters https://devblogs.microsoft.com/semantic-kernel/filters-in-semantic-kernel/ , the problem is knowing about the concept first.

I'm giving this feedback because I know that when you are into a project you live in a bubble that prevents you from seeing how "strangers" see through it.
I support this project because would help me alleviate writing "boilerplate stuff" and focus more on building the product. I personally created something similar at the same time when sk announced but I happily dropped it once I saw sk become more mature because I'll have less code to maintain, so take my feedback as a productive feedback!

P.S.
https://learn.microsoft.com/en-us/dotnet/api/microsoft.semantickernel.kernel.data?view=semantic-kernel-dotnet#microsoft-semantickernel-kernel-data this is the kind of "doc" that I personally find quite meaningless, it gives zero information more than what you get with intellisense, what gives value are more in-depth docs like lifecycle, internals (I'd like a deep dive on how autofunction works) and snippets (with advanced scenarios too)
